Add dither option to paint

DomiR opened the pull request:
I've added a gradient example in one commit that should not be merged and is only for testing.

This PR is an attempt to fix: #1368

wcandillon commented on example/src/App.tsx:
```diff
@@ -95,6 +97,14 @@ const App = () => {
               title: "🎨 Skia",
             }}
           />
+          <Stack.Screen
```
We don't need an extra example here, we can just add a gradient here: https://github.com/Shopify/react-native-skia/blob/main/example/src/Examples/API/Gradients.tsx

DomiR replied:
It's not very visible on with the small tile but I will remove this from PR as soon as dither is added, just wanted to show the problem...

wcandillon replied:
ok. Now that we have the test we can remove it.

wcandillon commented on package/src/dom/nodes/PaintNode.ts:
```diff
@@ -55,6 +56,10 @@ export class PaintNode
       paint.setAntiAlias(antiAlias);
     }
+    if (dither !== undefined) {
+      console.log('set dither', dither)
```
we can remove the log here

wcandillon commented on package/src/skia/web/JsiSkPaint.ts:
```diff
@@ -66,6 +66,10 @@ export class JsiSkPaint extends HostObject<Paint, "Paint"> implements SkPaint {
   }
 
+  setDither(aa: boolean) {
+    this.ref.setDither(aa);
```
setDither is not exposed in canvas kit, could that be?

DomiR replied:
Yes, see comment below. We would need to expose it...

DomiR commented:
Thanks for reviewing this. For SkiaKit to support dithering we would need to add it to the CanvasKit binding. 

https://github.com/google/skia/blob/935e20f3d1a53b7e390e9aaf513b702b92a9fc22/modules/canvaskit/canvaskit_bindings.cpp#L1660

Not sure if we should just open a ticket with skia or if you want to patch and build CanvasKit from source as well.

As I'm only interested in the react-native part I would be fine to make it noop for web temporarily.

So I cannot seem to figure out how the props on <Fill> are passed through. First I wanted to see if the flag would have an effect. It does as you can see below. I was able to `context->getPaint()->setDither(true);` in the JsiFillNode.h to make it work. But I cannot see how <Fill antiAlias> will pass the `antiAlias` to the context Paint. I just enabled dither in `JsiPaintNode.h`  but that did not work. In the example I've enabled dither to `DrawingContext.cpp` and that did the trick. Now I'm just not sure how the Fill props are beeing used. Maybe you can point me in the right direction. I have not checked yet, but adding a <Paint> would probably use `JsiPaintNode.h` and do the right thing, but then `<Fill antiAlias>` prop is somehow ignored?!

Maybe you can point me in the right direction here and show me how the `antiAlias` is passed to the jsi node for the Fill component.

Without dither             |  With dither
:-------------------------:|:-------------------------:
![screencap-2023-02-18-18 04 23](https://user-images.githubusercontent.com/1834664/219878764-df3d23ba-1ab9-478c-949a-d067b75f682f.png) |  ![screencap-2023-02-18-18 05 13](https://user-images.githubusercontent.com/1834664/219878766-4fa6540d-1a93-4caf-af5d-b58e79934a6e.png)

wcandillon commented on package/cpp/rnskia/dom/nodes/JsiFillNode.h:
```diff
@@ -15,6 +15,7 @@ class JsiFillNode : public JsiDomDrawingNode,
 
 protected:
   void draw(DrawingContext *context) override {
+    //context->getPaint()->setDither(true);
```
to remove

wcandillon commented on package/cpp/rnskia/dom/nodes/JsiPaintNode.h:
```diff
@@ -16,6 +16,7 @@ class JsiPaintNode : public JsiDomDeclarationNode,
   void decorate(DeclarationContext *context) override {
     auto paint = std::make_shared<SkPaint>();
     paint->setAntiAlias(true);
+    // paint->setDither(true);
```
to remove

wcandillon commented on package/src/skia/web/JsiSkPaint.ts:
```diff
@@ -66,6 +66,10 @@ export class JsiSkPaint extends HostObject<Paint, "Paint"> implements SkPaint {
   }
 
+  setDither(aa: boolean) {
+    // this.ref.setDither(aa);
```
we should throw a warning here

DomiR replied:
It's now available in SkiaWASM so we can keep it here.

wcandillon commented:
Looking good! Could you update the documentation too as mentioned above?

I change the target from main to integration so we can run the CI there and we will be able to put final touches before shipping it.
As you might have guessed already, in the Fill node, all paint property are handled automatically. I see that there is some handling missing in a couple of places. 
Please check the following files:
`src/dom/types/DrawingContext.ts`
`cpp/rnskia/dom/base/ConcatablePaint.cpp`
`cpp/rnskia/dom/nodes/JsiPaintNode.h`

DomiR commented:
Hey, sorry for the late reply.

So I think the main issue I see right now, is that the added test is a bit useless because we run it in node, meaning we are using skia wasm, where we would shim the setDither (does nothing). So adding a test with enabled and disabled would both give us the same result.

I've already asked whether or not you would be open to building skia wasm from source (as you already do this with android and ios). We would need to apply this patch first: [scripts/skia-wasm.patch](https://github.com/Shopify/react-native-skia/blob/55fa4b1764f8a9c157c4a268b4d0cbfb633ab0fa/scripts/skia-wasm.patch) which exposes setDither for wasm.

I tried this, and it seems to work fine, the resulting snapshot with dither enabled would look like this: [package/src/__tests__/snapshots/paint/dither.png](https://github.com/Shopify/react-native-skia/blob/55fa4b1764f8a9c157c4a268b4d0cbfb633ab0fa/package/src/__tests__/snapshots/paint/dither.png)

So before I clean up this PR I need to know if you would be open to building skia wasm from source with the applied patch. Alternatively I can request the patch upstream, but google created some hurdles for reporting skia issues/PRs.

One other thing - I would like so see a setAntiAlias test, as there does not seem to be a specific test for that yet. Would you be so kind to add one, so I can align my test to your standards? Or can you modify my test [here](https://github.com/Shopify/react-native-skia/blob/55fa4b1764f8a9c157c4a268b4d0cbfb633ab0fa/package/src/renderer/__tests__/e2e/Paint.spec.tsx#L94) not to use <Fill> but <Paint> directly setting antiAlias as a prop, so I would hit a debug breakpoint [here](https://github.com/Shopify/react-native-skia/blob/d1e2577db8a5a1b33fbd139fea027d529c6bce1f/package/src/skia/web/JsiSkPaint.ts#L66).

Gameplan:
- [ ] add skia wasm to build pipeline (and ultimately also in the npm package)
- [ ] add snapshot-test for both enabled and disabled dithering
- [ ] add documentation for the flag
- [ ] fix filled node to respect the dither prop

Let me know what you think!

wcandillon commented:
@DomiR Tests on the e2e folder are tested against Android and iOS, you can see an example at https://github.com/Shopify/react-native-skia/blob/main/package/src/renderer/__tests__/e2e/Mask.spec.tsx
To run the tests, run in `package/` `yarn e2e`, and to select a particular test `E2E=true yarn test -i e2e/TestName`. Now for these tests to run, you need to start the example project and open the screen "E2E Tests" from the menu. Let me know if you have any questions there.

About submitting a canvaskit patch, this is always a good idea. if you build skia locally (using these instructions: https://github.com/shopify/react-native-skia#react-native-skia), you can modify and build canvaskit in the local folder: (externals/skia/modules/canvaskit)) and use `ninja build` in the folder to build canvaskit. 

I hope this helps.

DomiR commented:
Hey @wcandillon, I've created an issue with skia and proposed the patch and they merged it pretty fast. But it took quite a while for them to publish the updated npm package, which they now did. I've already updated my PR on my machine last week, but was not able to complete it before I went on holiday 😅 I'll have to come back to this next month 🙈

wcandillon commented:
Dom, if you make it, we will merge it :)

DomiR commented:
Hey @wcandillon, sorry again for the long delay. So I've updated the docs and cleaned up the PR and resolved all your comments. I've rebased onto main as the `integration` branch seems to be a bit far behind, feel free to change this back.  

On my mac the e2e tests unfortunately won't run. Jest just hangs at `Determining test suites to run...` while the example app is running with e2e screen open as per the instructions. If there's anything else I can try please let me know.

When I run the test locally (via canvaskit-wasm), neither the new dither nor the existing antiAlias seem to be respected. `antiAlias` seems to be set to true every time: https://github.com/Shopify/react-native-skia/blob/a69ddc83b5c19f3cc39db73fdb81fba57a9f5bb3/package/src/skia/web/JsiSkia.ts#L67. As canvaskit-wasm now also has the option to set this, we could support this for web also.

Let me know what else you need for this to get merged. Best, Domi

wcandillon commented:
Looks good, I will merge this into an integration branch so we can check the CI and finalize everything. thanks a lot for your hard work!
